src/cephadm/cephadm

```diff
@@ -2728,10 +2728,50 @@ def create_daemon_dirs(ctx, fsid, daemon_type, daemon_id, uid, gid,
         sg = SNMPGateway.init(ctx, fsid, daemon_id)
         sg.create_daemon_conf()
 
+    _write_custom_conf_files(ctx, daemon_type, str(daemon_id), fsid, uid, gid)
+
+
+def _write_custom_conf_files(ctx: CephadmContext, daemon_type: str, daemon_id: str, fsid: str, uid: int, gid: int) -> None:
+    # mostly making this its own function to make unit testing easier
+    if 'config_json' in ctx and ctx.config_json:
+        config_json = get_custom_config_files(ctx.config_json)
+        if not os.path.exists(os.path.join(ctx.data_dir, fsid, 'custom_config_files', f'{daemon_type}.{daemon_id}')):
+            makedirs(os.path.join(ctx.data_dir, fsid, 'custom_config_files', f'{daemon_type}.{daemon_id}'), uid, gid, 0o755)
+        for ccf in config_json['custom_config_files']:
+            fname = ccf['filename']
+            if isinstance(fname, str):
+                with open(os.path.join(ctx.data_dir, fsid, 'custom_config_files', f'{daemon_type}.{daemon_id}', fname), 'w+', encoding='utf-8') as f:
+                    os.fchown(f.fileno(), uid, gid)
+                    os.fchmod(f.fileno(), 0o600)
+                    custom_content: str = ''
+                    if isinstance(ccf['content'], list):
+                        custom_content = '\n'.join(ccf['content'])
```

Comment:
I'm not sure we should prefer to assume line oriented content. While we're calling the feature custom config files, I assume there's nothing stopping anyone from writing custom data files, or even any other kind of file. See my review summary comment for more context.